algorithms/nautilus/grammartec/tree.cpp

```diff
@@ -0,0 +1,618 @@
+ * @author Ricerca Security <fuzzuf-dev@ricsec.co.jp>
+ *
+ * @details This file defines the tree and its operations.
+ *          There are two types of trees: Tree and TreeMutation,
+ *          and both inherits theTreeLike class.
+ */
+#include <unordered_set>
+#include "fuzzuf/algorithms/nautilus/grammartec/recursion_info.hpp"
+#include "fuzzuf/algorithms/nautilus/grammartec/tree.hpp"
+#include "fuzzuf/exceptions.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+
+namespace fuzzuf::algorithm::nautilus::grammartec {
+
+/**
+ * @fn
+ * Construct a tree from a vector of rules.
+ * @brief Construct a Tree instance
+ * @param (rules) Vector of rules
+ * @param (ctx) Context
+ */
+Tree::Tree(std::vector<RuleIDOrCustom>&& rules, Context& ctx)
+  : _rules(std::move(rules)) {
+  /* resize: Not only allocate buffer but also change size */
+  _sizes.resize(_rules.size(), 0);
+  _paren.resize(_rules.size(), NodeID(0));
+
+  if (_rules.size() > 0)
+    CalcSubTreeSizesAndParents(ctx);
+}
+
+/**
+ * @fn
+ * Calculate the sizes for each subtree and parent.
+ * @brief Calculate size information
+ * @param (ctx) Context
+ */
+void Tree::CalcSubTreeSizesAndParents(Context& ctx) {
+  CalcParents(ctx);
+  CalcSizes();
+}
+
+/**
+ * @fn
+ * Calculate parent information.
+ * @brief Calculate parents
+ * @param (ctx) Context
+ */
+void Tree::CalcParents(Context& ctx) {
+  if (Size() == 0)
+    return;
+
+  std::vector<std::pair<NTermID, NodeID>> stack;
+  stack.emplace_back(GetRule(NodeID(0), ctx).Nonterm(), NodeID(0));
+
+  for (size_t i = 0; i < Size(); i++) {
+    NodeID node_id(i);
+    const NTermID& nonterm = GetRule(node_id, ctx).Nonterm();
+
+    if (stack.size() == 0)
+      throw exceptions::fuzzuf_runtime_error(
+        "Not a valid tree for unparsing!", __FILE__, __LINE__
+      );
+
+    auto [nterm_id, node] = stack.back();
+    stack.pop_back();
+    if (nterm_id != nonterm)
+      throw exceptions::fuzzuf_runtime_error(
+        "Not a valid tree for unparsing!", __FILE__, __LINE__
+      );
+
+    _paren[i] = node;
+
+    const Rule& rule = GetRule(node_id, ctx);
+    std::vector<NTermID> nonterms = rule.Nonterms();
+    for (auto it = nonterms.crbegin(); it != nonterms.crend(); ++it) {
+      stack.emplace_back(*it, node_id);
+    }
+  }
+}
+
+/**
+ * @fn
+ * Calculate size information.
+ * @brief Calculate sizes
+ */
+void Tree::CalcSizes() {
+  for (size_t& size: _sizes)
+    size = 1;
+
+  for (size_t i = Size() - 1; i > 0; i--)
+    _sizes.at(static_cast<size_t>(_paren.at(i))) += _sizes[i];
+}
+
+/**
+ * @fn
+ * @brief Get the slice of rules by node IDs.
+ * @param (from) Start node ID of slice
+ * @param (to) End node ID of slice
+ * @return Vector of RuleIDOrCustom derived from slice
+ */
+std::vector<RuleIDOrCustom> Tree::Slice(
+  const NodeID& from, const NodeID& to
+) const {
+  return std::vector<RuleIDOrCustom> (
+    _rules.begin() + static_cast<size_t>(from), 
+    _rules.begin() + static_cast<size_t>(to)
+  );
+}
```

Comment:
Unlike Rust's slice, this causes a copy. Is that no problem?

Reply:
This will be left as FIXME in [4e67411](https://github.com/fuzzuf/fuzzuf/pull/23/commits/4e6741106f34cc8f582ec4c992ee17be46d30441) because the size of `RuleIDOrCustom` and the slice is relatively small.

Reply:
Got it.